File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/vehicle/GroundReactionForceCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator.vehicle;

import com.github.aadvorak.artilleryonline.battle.calculations.*;
import com.github.aadvorak.artilleryonline.battle.calculator.utils.ForceCalculatorUtils;
import com.github.aadvorak.artilleryonline.battle.common.Constants;
import com.github.aadvorak.artilleryonline.battle.common.Force;
import com.github.aadvorak.artilleryonline.battle.config.VehicleConfig;
import com.github.aadvorak.artilleryonline.battle.model.BattleModel;
import com.github.aadvorak.artilleryonline.battle.model.VehicleModel;
import com.github.aadvorak.artilleryonline.battle.precalc.VehiclePreCalc;
import com.github.aadvorak.artilleryonline.battle.specs.VehicleSpecs;
import com.github.aadvorak.artilleryonline.battle.state.VehicleState;

import java.util.ArrayList;
import java.util.List;

public class GroundReactionForceCalculator implements ForceCalculator<
        VehicleSpecs,
        VehiclePreCalc,
        VehicleConfig,
        VehicleState,
        VehicleModel,
        VehicleCalculations> {

    private static final String FORCE_DESCRIPTION = "Ground reaction";

    @Override
    public List<BodyForce> calculate(VehicleCalculations calculations, BattleModel battleModel) {
        var forces = new ArrayList<BodyForce>();
        if (calculations.getModel().getState().getVelocity().magnitude() < Constants.ZERO_THRESHOLD) {
            return forces;
        }
        var groundReactionCoefficient = battleModel.getRoom().getSpecs().getGroundReactionCoefficient();
        var groundMaxDepth = battleModel.getRoom().getSpecs().getGroundMaxDepth();
        addWheelReaction(forces, calculations.getRightWheel(), groundReactionCoefficient, groundMaxDepth);
        addWheelReaction(forces, calculations.getLeftWheel(), groundReactionCoefficient, groundMaxDepth);
        addTurretReaction(forces, calculations, groundReactionCoefficient, groundMaxDepth);
        return forces;
    }

    private void addWheelReaction(List<BodyForce> forces, WheelCalculations calculations,
                                  double groundReactionCoefficient, double groundMaxDepth) {
        if (calculations.getGroundContact() == null) {
            return;
        }
        var velocityNormalProjection = calculations.getVelocity()
                .projectionOnto(calculations.getGroundContact().normal());
        var velocityNormalProjectionMagnitude = velocityNormalProjection.magnitude();
        if (velocityNormalProjectionMagnitude > Constants.ZERO_THRESHOLD
                && velocityNormalProjection.dotProduct(calculations.getGroundContact().normal()) > 0) {
            var depth = Math.min(calculations.getGroundContact().depth(), groundMaxDepth);
            var force = Force.of(calculations.getGroundContact().normal()
                    .multiply(- velocityNormalProjectionMagnitude * depth * groundReactionCoefficient));
            forces.add(BodyForce.of(force, calculations.getGroundContact().position(),
                    calculations.getModel().getState().getPosition().getCenter(), FORCE_DESCRIPTION));
        }
    }

    private void addTurretReaction(List<BodyForce> forces, VehicleCalculations calculations,
                                   double groundReactionCoefficient, double groundMaxDepth) {
        if (calculations.getTurretGroundContacts() == null) {
            return;
        }
        calculations.getTurretGroundContacts().forEach(contact -> ForceCalculatorUtils.addReaction(forces,
                calculations, contact, groundMaxDepth, groundReactionCoefficient, FORCE_DESCRIPTION));
    }
}

File: frontend/playground/battle/calculator/common/ground-reaction-force-calculator.ts
import type {ForceCalculator} from "~/playground/battle/calculator/force-calculator";
import type {BodyCalculations} from "~/playground/data/calculations";
import type {BattleModel} from "~/playground/data/model";
import {BodyForce} from "~/playground/battle/calculator/body-force";
import {VectorUtils} from "~/playground/utils/vector-utils";
import {BodyUtils} from "~/playground/utils/body-utils";
import {Constants} from "~/playground/data/constants";

export class GroundReactionForceCalculator implements ForceCalculator<BodyCalculations> {
  private static readonly FORCE_DESCRIPTION = 'Ground reaction'

  calculate(calculations: BodyCalculations, battleModel: BattleModel): BodyForce[] {
    const forces: BodyForce[] = []
    if (VectorUtils.getBodyMagnitude(calculations.model.state.velocity) < Constants.ZERO_THRESHOLD) {
      return forces
    }
    const groundReactionCoefficient = battleModel.room.specs.groundReactionCoefficient
    const groundMaxDepth = battleModel.room.specs.groundMaxDepth
    this.addReaction(forces, calculations, groundReactionCoefficient, groundMaxDepth)
    return forces
  }

  private addReaction(
      forces: BodyForce[],
      calculations: BodyCalculations,
      groundReactionCoefficient: number,
      groundMaxDepth: number
  ): void {
    const groundContacts = calculations.getGroundContacts()
    if (!groundContacts) {
      return
    }
    const comPosition = calculations.model.state.position

    groundContacts.forEach(contact => {
      const velocityNormalProjection = VectorUtils.projectionOfOnto(
          BodyUtils.getVelocityAt(calculations.model.state, contact.position), contact.normal)
      const velocityNormalProjectionMagnitude = VectorUtils.getMagnitude(velocityNormalProjection)
      if (velocityNormalProjectionMagnitude > Constants.ZERO_THRESHOLD
          && VectorUtils.dotProduct(velocityNormalProjection, contact.normal) > 0) {
        const depth = Math.min(contact.depth, groundMaxDepth)
        const force = VectorUtils.multiply(contact.normal,
            -velocityNormalProjectionMagnitude * depth * groundReactionCoefficient)

        forces.push(BodyForce.of(
            force,
            contact.position,
            comPosition,
            GroundReactionForceCalculator.FORCE_DESCRIPTION
        ))
      }
    })
  }
}
